src/main/java/com/wiley/autotest/selenium/elements/upgrade/BaseTeasyElement.java

```diff
@@ -0,0 +1,495 @@
+package com.wiley.autotest.selenium.elements.upgrade;
+
+import com.wiley.autotest.selenium.Report;
+import com.wiley.autotest.selenium.context.SearchStrategy;
+import com.wiley.autotest.selenium.driver.FramesTransparentWebDriver;
+import com.wiley.autotest.selenium.elements.upgrade.v3.TeasyElementFinder;
+import org.openqa.selenium.*;
+import org.openqa.selenium.interactions.Actions;
+import org.openqa.selenium.interactions.internal.Coordinates;
+import org.openqa.selenium.internal.Locatable;
+import org.openqa.selenium.remote.Augmenter;
+import org.openqa.selenium.remote.RemoteWebDriver;
+import org.openqa.selenium.remote.UnreachableBrowserException;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.UndeclaredThrowableException;
+import java.util.List;
+
+import static com.wiley.autotest.selenium.SeleniumHolder.getWebDriver;
+import static com.wiley.autotest.utils.ExecutionUtils.isChrome;
+import static com.wiley.autotest.utils.TestUtils.fail;
+import static com.wiley.autotest.utils.TestUtils.waitForSomeTime;
+
+
+/**
+ * User: vefimov
+ * Date: 27.08.2014
+ * Time: 15:41
+ */
+public abstract class BaseTeasyElement implements TeasyElement, Locatable {
+
+    private WebElement wrappedElement;
+    private Locator locator;
+    private TeasyElementFinder contextFinder;
+
+    //TODO think about making this value configurable
+    private static final long TIMEOUT_FOR_AGAIN_LOCATE_IN_SECONDS = 5;
+    private int repeatLocateElementCounter;
+    private static final int MAX_NUMBER_OF_REPEAT_LOCATE_ELEMENT = 20;
+
+    BaseTeasyElement(TeasyElementData teasyElementData) {
+        WebElement element = teasyElementData.getElement();
+        TeasyElement searchContext = teasyElementData.getSearchContext();
+        By by = teasyElementData.getBy();
+        Integer index = teasyElementData.getIndex();
+
+        if (searchContext != null && by != null && index != null) {
+            //element from list in searchContext with index
+            this.locator = new FindElementsLocator(searchContext, by, index);
+        } else if (by != null && index != null) {
+            //element from list with index
+            this.locator = new FindElementsLocator(getWebDriver(), by, index);
+        } else if (searchContext != null && by != null) {
```

Comment:
Should be extracted to ElementsLocatorFactory class

Reply:
@dpavlov good point will do it